File: AOC2017/AOC2017.hs
import Data.Char
import Data.List
import Data.Maybe
import Debug.Trace
import Text.RegexPR
import Control.Applicative
import Text.ParserCombinators.ReadP
import qualified Data.HashMap.Lazy as HL
import qualified Data.Array.Unboxed as DA
import qualified Data.HashMap.Strict as H

data Part = PartA | PartB deriving Eq

parsePRGroups :: Int -> [(Int, String)] -> [Maybe String]
parsePRGroups n groups = [fmap snd $ find ((== i) . fst) groups | i <- [1 .. n]]

day01 :: Part -> String -> Int
day01 part input
  = sum $ map (digitToInt.fst) $ filter (uncurry (==)) pairs
  where pairs = case part of PartA -> zip digits $ tail digits ++ [head digits]
                             PartB -> zip digits $ drop (length digits `div` 2) digits ++ digits
        digits = filter isDigit input

day02 :: Part -> String -> Int
day02 part input
  = sum $ map (lineFn . map read . words) $ lines input
  where lineFn r = case part of PartA -> maximum r - minimum r
                                PartB -> head [n `div` d | n <- r, d <- r, n /= d, n `mod` d == 0]

day03 :: Part -> Int -> Int
day03 PartA input =
  let major_coord = ceiling $ fromIntegral input ** 0.5 * 0.5 - 0.5
      minor_diag  = ((2 * major_coord + 1) ^ 2 - input) `mod` (2 * major_coord)
      minor_coord = abs $ major_coord - minor_diag
  in  minor_coord + major_coord

day03 PartB input = iter (H.singleton (0, 0) 1) (0, 0)
  where iter :: H.HashMap (Int, Int) Int -> (Int, Int) -> Int
        iter hsh (x, y) | hsh H.! (x, y) > input = hsh H.! (x, y)
                        | otherwise = iter (nextHsh hsh $ nextCoord (x, y)) $ nextCoord (x, y)
        nextHsh hsh (nx, ny) = H.insert (nx, ny) (nextValue hsh (nx, ny)) hsh
        nextValue hsh (nx, ny) = sum [H.lookupDefault 0 (nx + dx, ny + dy) hsh
                                     | dx <- [-1, 0, 1], dy <- [-1, 0, 1]]
        nextCoord (x, y) | x ==  y && y < 0 = (x, y + 1)
                         | x == -y && y < 0 = (x - 1, y)
                         | abs x == y       = (x + 1, y)
                         | abs x < abs y    = (x + signum y, y)
                         | otherwise        = (x, y - signum x)

day04 :: Part -> String -> Int
day04 part input = length [() | l <- lines input, 
                                let ws = (case part of PartA -> id; PartB -> map sort) $ words l,
                                ws == nub ws]

day05 :: Part -> String -> Int
day05 part input = iter 0 0 $ DA.listArray (0, length inArray - 1) inArray
  where inArray = map read $ words input :: [Int]
        iter :: Int -> Int -> DA.Array Int Int -> Int
        iter t ix ary | not $ DA.inRange (DA.bounds ary) ix = t
        iter t ix ary = iter (t+1) (ix + ary DA.! ix) (ary DA.// [(ix, nextOffset $ ary DA.! ix)])
        nextOffset offset | offset >= 3 && part == PartB = offset - 1
                          | otherwise = offset + 1

day06 :: String -> (Int, Int)
day06 input = iter 0 (map read $ words input) H.empty
  where iter :: Int -> [Int] -> H.HashMap [Int] Int -> (Int, Int)
        iter t state seen | H.member state seen = (t, t - seen H.! state)
        iter t state seen = iter (t+1) (nextState state) (H.insert state t seen)
        nextState :: [Int] -> [Int]
        nextState state = 
          let stateSize = length state
              maxValue = maximum state
              Just maxIndex = elemIndex maxValue state
              in  [ (if index == maxIndex then 0 else el) 
                  + (maxValue `div` stateSize)
                  + (if (index - maxIndex - 1) `mod` stateSize < maxValue `mod` stateSize then 1 else 0)
                  | (el, index) <- zip state [0..]]

data Day07Node = Day07Node String Int [Day07Node] deriving (Eq, Show)
day07 :: Part -> String -> Either String Int
day07 part input = let nodesByName :: H.HashMap String Day07Node
                       nodesByName = H.fromList [(name, Day07Node name (read weightStr) children)
                                                | (_, groups) <- gmatchRegexPR "([a-z]*) \\((\\d+)\\)(?: -> ([a-z, ]+))?" input,
                                                  let [Just name, Just weightStr, maybeChildStr] = parsePRGroups 3 groups,
                                                  let children = map (nodesByName H.!) $ splitRegexPR ", " $ fromMaybe "" maybeChildStr]
                       [root] = HL.elems nodesByName \\ [child | Day07Node _ _ children <- H.elems nodesByName, child <- children]
                       treeWeight :: Day07Node -> Int
                       treeWeight (Day07Node _ weight children) = weight + sum (map treeWeight children)
                       unbalancedNodes = [(node, correctWeight - wrongWeight + correctNodeWeight)
                                         | node@(Day07Node name weight children) <- H.elems nodesByName,
                                           let childWeights = map treeWeight children,
                                           (>= 2) $ length $ nub childWeights,
                                           let ([correctWeight:_], [[wrongWeight]]) = partition ((>1).length) $ group $ sort childWeights,
                                           let [correctNodeWeight] = [weight | node@(Day07Node _ weight _) <- children, treeWeight node == wrongWeight]
                                         ]
                       [fixedWeight] = [fixedWeight 
                                       | (Day07Node _ _ children, fixedWeight) <- unbalancedNodes,
                                         null $ intersect children $ map fst unbalancedNodes]
                   in  case part of PartA -> case root of Day07Node name _ _ -> Left name
                                    PartB -> Right fixedWeight

day08 :: String -> (Int, Int)
day08 input = case foldl' iter (0, H.empty) (lines input) 
              of   (submax, regs) -> (maximum $ H.elems regs, submax)
  where iter :: (Int, H.HashMap String Int) -> String -> (Int, H.HashMap String Int)
        iter (submax, regs) line =
          let Just (_, groups) = matchRegexPR "([a-z]+) (inc|dec) (-?\\d+) if ([a-z]+) (>|>=|==|!=|<|<=) (-?\\d+)" line
              [Just regWr, Just opWr, Just nrWr, Just regCond, Just opCond, Just nrCond] = parsePRGroups 6 groups
              newHash = if testCond opCond (H.lookupDefault 0 regCond regs) (read nrCond) 
                        then H.insert regWr (fn opWr (H.lookupDefault 0 regWr regs) (read nrWr)) regs
                        else regs
              fn :: String -> Int -> Int -> Int; fn "dec" = (-); fn "inc" = (+)
              testCond "<" = (<); testCond "<=" = (<=); testCond "==" = (==)
              testCond "!=" = (/=); testCond ">" = (>); testCond ">=" = (>=)
              
          in  (maximum $ submax : H.elems newHash, newHash)

data Day09Node = Day09Stream [Day09Node] | Day09Garbage Int deriving Show
  
day09 :: String -> (Int, Int)
day09 input =
  let parseNode :: ReadP Day09Node
      parseNode = parseStream <|> parseGarbage
      parseStream = fmap Day09Stream $ between (char '{') (char '}') $ parseNode `sepBy` (char ',')
      parseGarbage = fmap (Day09Garbage . sum) $ (char '<' >>) $ flip manyTill (char '>') $
        (char '!' >> get >> return 0) <++ (get >> return 1)
      [(stream, _)] = readP_to_S parseStream input
        
      score :: Int -> Day09Node -> Int
      score depth (Day09Stream elems) = depth + (sum $ map (score $ depth + 1) elems)
      score _ (Day09Garbage _) = 0
      
      garbage :: Day09Node -> Int
      garbage (Day09Stream elems) = sum $ map garbage elems
      garbage (Day09Garbage n) = n
  in  (score 1 stream, garbage stream)

main = print . day09 =<< readFile "day09in.txt"
-- main = print $ day03 PartB 325489
-- main = print $ day06 "4 10 4 1 8 4 9 14 5 1 14 15 0 15 3 5"

File: AOC2017/AOC2017.rb
def day01(input, part)
  input.tr! "^0-9", ""
  pairs = case part
  when :a then (input + input[0]).chars.each_cons(2)
  when :b then input[0 ... input.size/2].chars.zip(input[input.size/2 .. -1].chars) * 2
  end
  pairs.select{|a, b| a==b}.map{|a, b| a.to_i}.reduce(0, &:+)
end

def day02(input, part)
  rows = input.lines.map{|r| r.split(" ").map(&:to_i)}
  case part
  when :a then rows.map{|r| r.max - r.min}.reduce(0, &:+)
  when :b then rows.map{|r| x = nil; r.find{|n| r.find{|d| x = n / d if n != d && n % d == 0}}; x}.reduce(0, &:+)
  end
end

def day03(input, part)
  case part
  when :a
    major_coord = (input ** 0.5 * 0.5 - 0.5).ceil
    minor_coord = ((2 * major_coord + 1) ** 2 - input) % (2 * major_coord)
    minor_coord = (major_coord - minor_coord).abs
    
    minor_coord + major_coord
  when :b
    board = Hash.new(0); board[[0, 0]] = 1; x = 0; y = 0
    loop.with_index(1) do
      p [x, y, board[[x, y]]]; gets
      return board[[x, y]] if board[[x, y]] > input
      
      case
      when x == y && x < 0 then y += 1
      when x.abs == y.abs then x += y <=> -0.5
      when x.abs < y.abs then x += y <=> 0
      else y += 0 <=> x
      end

      board[[x, y]] = [[0, 1], [0, -1], [1, 0], [-1, 0], [-1, -1], [-1, 1], [1, -1], [1, 1]]
          .map{|dx, dy| board[[x+dx, y+dy]]}.reduce(&:+)
    end
  end
end

def day04(input, part)
  input.lines.count do |line|
    ws = line.split(" ")
    ws.map!{|w| w.chars.sort.join} if part == :b
    ws == ws.uniq
  end
end

def day05(input, part)
  maze = input.lines.map(&:to_i)
  maze_ix = 0
  loop.with_index(1) do |_, t|
    old_ix = maze_ix
    maze_ix += maze[maze_ix]
    maze[old_ix] += part == :b && maze[old_ix] >= 3 ? -1 : 1
    return t if maze_ix < 0 || maze_ix >= maze.size
  end
end

day06in = "4	10	4	1	8	4	9	14	5	1	14	15	0	15	3	5"

def day06(input, part)
  state = input.split("\t").map(&:to_i)
  seen = {}
  loop.with_index do |_, t|
    return  t - seen[state], t if seen[state]
    seen[state] = t
    max_val = state.max
    max_ix = state.find_index max_val
    state = state.map.with_index{|e, i| (max_ix == i ? 0 : e) + max_val / state.size +
                                        ((i - max_ix - 1) % state.size < max_val % state.size ? 1 : 0)}
  end
end

Day07Node = Struct.new :name, :weight, :child_names, :children
def day07(input, part)
  nodes = input.lines.map do |line|
    line =~ /([a-z]+) \((\d+)\)(?: -> ([a-z, ]+))?/
    Day07Node.new $1, $2.to_i, $3 ? $3.split(", ") : [], nil
  end
  
  root_candidates = nodes.dup
  nodes_by_name = Hash[nodes.map{|node| [node.name, node]}]
  nodes.each{|node| node.children = node.child_names.map{|name| nodes_by_name[name]}}
  nodes.each{|node| node.children.each{|c| root_candidates.delete c}}
  
  root = root_candidates[0]
  return root_name if part == :a  
  
  subtree_weight_or_return = Proc.new do |node|
    weights = node.children.map(&subtree_weight_or_return)
    if weights.uniq.size <= 1
      weights.reduce(0, &:+) + node.weight
    else
      wrong_weight_at = weights.find_index{|w| weights.count(w) == 1}
      wrong_weight = weights[wrong_weight_at]
      correct_weight = weights[wrong_weight_at == 0 ? 1 : 0]
      return correct_weight - wrong_weight + node.children[wrong_weight_at].weight
    end
  end
  subtree_weight_or_return[root]
end

def day08(input, part)
  regs = Hash.new(0)
  max = 0
  input.lines.each do |line|
    line =~ /([a-z]+) (inc|dec) (-?\d+) if ([a-z]+) (>|>=|==|!=|<|<=) (-?\d+)/
    regs[$1] += $3.to_i * ($2 == "inc" ? 1 : -1) if regs[$4].send($5, $6.to_i)
    max = [max, regs.values.max].max
  end
  [regs.values.max, max]
end

class ParseError < StandardError; end
Day09Stream = Struct.new :elems, :score, :score_total, :garbage_removed do
    def self.parse_from(io, score)
      elems = []
      await_comma = false
      loop do
        case (char = io.read(1))
        when "{"
          raise ParseError if await_comma
          elems << Day09Stream.parse_from(io, score + 1)
          await_comma = true
        when "<" 
          raise ParseError if await_comma
          elems << Day09Garbage.parse_from(io)
          await_comma = true
        when ","
          raise ParseError unless await_comma
          await_comma = false
        when "}" 
          return new(elems, score, score + elems.map(&:score_total).reduce(0, &:+),
                     elems.map(&:garbage_removed).reduce(0, &:+))
        else raise ParseError
        end
      end
    end
end
Day09Garbage = Struct.new :garbage_removed do
  def score_total; 0; end
  def self.parse_from(io)
    garbage_removed = 0
    loop do
      case (char = io.read(1))
      when ">" then return new(garbage_removed)
      when "!" then io.read(1)
      else garbage_removed += 1
      end
    end
  end
end
def day09(input, part)
  io = StringIO.new(input)
  raise ParseError unless io.read(1) == "{"
  stream = Day09Stream.parse_from(io, 1)
  return stream.score_total, stream.garbage_removed
end

time = Time.now
p day09(File.read("day09in.txt"), :b)
puts "time: #{Time.now - time}"
